store/store.go

```diff
@@ -0,0 +1,519 @@
+	f, err = file.OpenQ1Q4File(path)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, ErrNotFound
+		}
+		return nil, fmt.Errorf("opening ODS file: %w", err)
+	}
+	return wrappedFile(f), nil
+}
+
+func (s *Store) HasByHash(ctx context.Context, datahash share.DataHash) (bool, error) {
+	if datahash.IsEmptyRoot() {
+		return true, nil
+	}
+	lock := s.stripLock.byDatahash(datahash)
+	lock.RLock()
+	defer lock.RUnlock()
+
+	tNow := time.Now()
+	exist, err := s.hasByHash(datahash)
+	s.metrics.observeHas(ctx, time.Since(tNow), err != nil)
+	return exist, err
+}
+
+func (s *Store) hasByHash(datahash share.DataHash) (bool, error) {
+	if datahash.IsEmptyRoot() {
+		return true, nil
+	}
+	path := s.basepath + blocksPath + datahash.String()
+	return pathExists(path)
+}
+
+func (s *Store) HasByHeight(ctx context.Context, height uint64) (bool, error) {
+	lock := s.stripLock.byHeight(height)
+	lock.RLock()
+	defer lock.RUnlock()
+
+	tNow := time.Now()
+	exist, err := s.hasByHeight(height)
+	s.metrics.observeHas(ctx, time.Since(tNow), err != nil)
+	return exist, err
+}
+
+func (s *Store) hasByHeight(height uint64) (bool, error) {
+	if s.isEmptyHeight(height) {
+		return true, nil
+	}
+
+	_, err := s.cache.Get(height)
+	if err == nil {
+		return true, nil
+	}
+
+	path := s.basepath + heightsPath + fmt.Sprintf("%d", height)
+	return pathExists(path)
+}
+
+func (s *Store) Remove(ctx context.Context, height uint64) error {
+	lock := s.stripLock.byHeight(height)
+	lock.Lock()
+	defer lock.Unlock()
+
+	tNow := time.Now()
+	err := s.remove(ctx, height)
+	s.metrics.observeRemove(ctx, time.Since(tNow), err != nil)
+	return err
+}
+
+func (s *Store) remove(ctx context.Context, height uint64) error {
+	f, err := s.getByHeight(height)
+	if err != nil {
+		// short circuit if file not exists
+		if errors.Is(err, ErrNotFound) {
+			return nil
+		}
+		return fmt.Errorf("getting by height: %w", err)
+	}
+
+	hash, err := f.DataHash(ctx)
+	if err != nil {
+		return fmt.Errorf("getting data hash: %w", err)
+	}
+	// close file to release the reference in the cache
+	if err = f.Close(); err != nil {
+		return fmt.Errorf("closing file on removal: %w", err)
+	}
+
+	// lock by datahash to prevent concurrent access to the same underlying file
+	// using links from different heights
+	dlock := s.stripLock.byDatahash(hash)
+	dlock.Lock()
+	defer dlock.Unlock()
+
+	if err = s.cache.Remove(height); err != nil {
+		return fmt.Errorf("removing from cache: %w", err)
+	}
+
+	// remove hard link by height
+	heightPath := s.basepath + heightsPath + fmt.Sprintf("%d", height)
+	if err = os.Remove(heightPath); err != nil {
+		return fmt.Errorf("removing by height: %w", err)
+	}
+
+	hashPath := s.basepath + blocksPath + hash.String()
+	count, err := linksCount(hashPath)
+	if err != nil {
+		return fmt.Errorf("counting links: %w", err)
+	}
+	if count == 1 {
+		err = os.Remove(hashPath)
+		if err != nil {
+			return fmt.Errorf("removing by hash: %w", err)
+		}
+	}
```

Comment:
We can simplify removal logic as that's impossible to make duplicate files unless its an empty file that we can handle differently